projs/shadow/shadow-engine/core/src/core/job/Job.cpp

```diff
@@ -0,0 +1,177 @@
+#include <cassert>
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <thread>
+#include <shadow/core/Time.h>
+#include <shadow/core/jobs/Job.h>
+
+#include "spdlog/spdlog.h"
+
+namespace SH::Jobs {
+    struct State {
+        size_t nCores = 0;
+        PriorityQueue priorities[Priority::SIZE];
+        std::atomic_bool alive { true };
+
+        void Destroy() {
+            alive.store(false);
+            bool wake = true;
+            // Notify all threads to finish their work
+            std::thread waker([&] { while (wake) for (auto& x : priorities) x.wake.notify_all(); });
+            // Wait for all threads to stop
+            for (auto& x : priorities) for (auto& thread : x.threads) thread.join();
+            wake = false;
+            // Wait for all threads' dependencies to finish
+            waker.join();
+
+            // Cleanup
+            for (auto& x : priorities) {
+                x.queues.reset();
+                x.threads.clear();
+                x.nThreads = 0;
+            }
+            nCores = 0;
+        }
+
+        ~State() { Destroy(); }
+    };
+
+    static State internalState;
+
+    void Init(size_t maxThreads) {
+        // Don't reinit if we're already doing something, that's BAD
+        if (internalState.nCores > 0) return;
```

Comment:
If you remove the list of running threads while a thread is running, it becomes impossible to join all those threads back, and the process CANNOT stop cleanly any more.

It feels weird because C++ threading is a fucking disaster. We need to just deal with it in this case.

Reply:
The comment is on the fact that you are using the `nCores` as an inite'd flag.
Like yes, there is little to no chance of this running on a computer with 0 cores. But still that field is now pulling double duty.